### PR TITLE
ci: cache refactor

### PR DESCRIPTION
**Pull request by lodmfjord**
# ...

Attach a link to issue if relevant

## What

Specify what you're trying to achieve

## Why

Specify why you need to achieve this

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
  - Introduced a new caching mechanism for CI/CD pipelines to improve build efficiency.
  - Added custom actions for cache retrieval and management in GitHub workflows.

- **Improvements**
  - Upgraded `actions/cache` and `actions/checkout` to the latest versions for better performance and security.
  - Restructured caching steps in workflows for more streamlined and reliable cache handling.

- **Chores**
  - Updated workflow files to reflect the new caching strategy and action versions.

### CODE REVIEW

**Comment by lodmfjord** [comment]
@coderabbitai resolve

File: .github/actions/unit-test/action.yml [hunk 1]
```diff
@@ -53,7 +53,7 @@ runs:
 
     - name: Cached codecov uploader
       id: codecov-cache
-      uses: actions/cache@v3
+      uses: actions/cache@v4
```
Comment (busla) [inline .github/actions/unit-test/action.yml:56]: nice 🤩

File: .github/actions/get-cache/action.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        script: |
+          core.exportVariable('ACTIONS_CACHE_URL', 'https://cache.dev01.devland.is/')
+          core.exportVariable('ACTIONS_RUNTIME_TOKEN', process.env['ACTIONS_RUNTIME_TOKEN'])
+          core.exportVariable('ACTIONS_RUNTIME_URL', process.env['ACTIONS_RUNTIME_URL'])
+    - name: Get cache
+      id: prepare
+      shell: bash
+      env:
+        _CACHE_KEYS: ${{ inputs.keys }}
+        ENABLE_CACHE: ${{ inputs.enable-cache }}
```
Comment (busla) [inline .github/actions/get-cache/action.yml:34]: I have a hard time tracking this env throughout the stack. Where is it used?

File: scripts/ci/cache/__config.mjs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    return a
+  }, {})
+
+export const cypressPath = '/github/home/.cache/Cypress'
+export const cacheSuccess = JSON.parse(process.env[ENV_CACHE_SUCCESS] ?? '{}')
+export const initCache = process.env[ENV_INIT_CACHE] === 'true'
+
+export const caches = [
+  {
+    enabled: ENABLED_MODULES['node_modules'],
```
Comment (busla) [inline scripts/ci/cache/__config.mjs:42]: do we have to hardcode a new object if we decide to add a new directory to cache?

File: scripts/ci/cache/_common.mjs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// @ts-check
+import { dirname, resolve } from 'path'
+import { fileURLToPath } from 'url'
+import { ENV_KEYS } from './_const.mjs'
+const __dirname = dirname(fileURLToPath(import.meta.url))
+export const SCRIPTS_DIR = __dirname
+export const ROOT = resolve(__dirname, '..', '..', '..')
```
Comment (busla) [inline scripts/ci/cache/_common.mjs:7]: could we try/catch/throw on this line?

File: scripts/ci/cache/_common.mjs [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+export const MOBILE_APP_DIR = resolve(ROOT, ...'apps/native/app'.split('/'))
```
Comment (busla) [inline scripts/ci/cache/_common.mjs:8]: could we try/catch/throw on this line?

File: scripts/ci/cache/_utils.mjs [hunk 6]
```diff
@@ -0,0 +1,218 @@
+// @ts-check
+import { arch, platform } from 'node:os'
+import crypto from 'node:crypto'
+import { ROOT } from './_common.mjs'
+import { spawn, exec } from 'node:child_process'
+import { resolve, join } from 'node:path'
+import { readFile, readdir, stat } from 'node:fs/promises'
+import { HASH_GENERATE_FILES_SCRIPT } from './_const.mjs'
+
+const GENERATE_HASH_GENERATED_FILES_SCRIPT = resolve(
```
Comment (busla) [inline scripts/ci/cache/_utils.mjs:10]: could we try/catch/throw on this line?

**Review by busla — APPROVED** [review]
LGTM

File: scripts/ci/cache/_utils.mjs (hunk 6 above)
Comment (lodmfjord) [reply scripts/ci/cache/_utils.mjs:10]: This only returns path, doesn't check if it exists.

File: scripts/ci/cache/__config.mjs (hunk 3 above)
Comment (lodmfjord) [reply scripts/ci/cache/__config.mjs:42]: no, path can also be an array

(which is what is supported by github actions cache)

File: .github/actions/get-cache/action.yml (hunk 2 above)
Comment (lodmfjord) [reply .github/actions/get-cache/action.yml:34]: _CACHE_KEYS is a string of json with all hash keys, only used if this is not inited.

ENABLE_CACHE enables or disables cache steps.

**Comment by lodmfjord** [comment]
@coderabbitai resolve

**Pull request merged**